### DISCUSSION TITLE
Using AT for non-relativistic particles

### DISCUSSION THREAD

**Discussion by lfarv** [atcollab/at#308] [Ideas]
I started looking at the possibility to use AT to track non-relativistic particles. It's rather easy to associate a particle to a lattice so that the relativistic &beta; and &gamma; are available where needed. However the tracking has a severe problem. With the choice of c&tau; and &Delta;p/p as longitudinal variables, it does not work. This choice implies &beta; = 1 and &Delta;p/p = &Delta;E/E. To make it work, we need to include some information on velocity in the particle coordinates. Looking around, I found two ways of doing this: the "classical" choice ('elegant', for instance), and the MAD-X choice.

1. **Classical coordinates**: the velocity information is included in the 6<sup>th</sup> coordinate which becomes &beta;c&tau; = &Delta;s instead of c&tau;. This does not make any difference in AT for all passmethods except cavities because:
  - the tracking through these elements does not depend upon the arrival time of the particle (6<sup>th</sup> coordinate),
  - What is computed there as c&tau; is indeed already &Delta;s. It's pure geometry. I checked `DriftPass` and the drift case of `ExactHamitonianPass` (where there is a bug, by the way): it's indeed &Delta;s.
  Only `CavityPass` has to be modified, because it computes an energy kick and not a momentum kick (they are identical for relativistic particles).
2. **MAD-X coordinates**: the velocity information is included in the 5<sup>th</sup> coordinate (6<sup>th</sup> in MAD-X), which is &Delta;E/pc = &beta;.&Delta;p/p. This cannot be envisaged since it implies rewriting the whole tracking.

To summarise, what is needed to have AT work for non-relativistic particles is rather easy, on the principle:
- Acknowledge the fact that the 6<sup>th</sup> coordinate is &Delta;s, **which it already is**, in reality,
- Modify `CavityPass` to introduce a 1/&beta;<sup>2</sup> factor to convert the &Delta;E/E energy kick into a &Delta;p/p kick.

This modification is small enough to minimise the risk of breaking anything, however, this is a critical point which has to be carefully checked before implementation. For instance, in usual electron machines, &beta; is not exactly equal to one, so it will induce a small change of the nominal RF or revolution frequency. Also, while most pass methods are based on drift/kick sequences, which are safe, there may be a few ones to be checked.

@lnadolski, @swhite2401, @simoneliuzzo, @carmignani, any ideas ?

**Comment by swhite2401** [comment]
Hello @lfarv, this is certainly an interesting development and I would even extend the discussion to arbitrary particles (protons, ions, etc...). I believe I saw the electron mass hard-coded in many places and it would probably be a good thing to clean this up as a first step.
Introducing general `at` variables to define the particle type (starting with electrons only) that all function refer to would be a first step in this direction. This would also be coherent with the suggestion made in #285 to introduce a `Constants.py` module to centralized physics constants.

For your specific problem, I would have to check the impedance passmethod, in its present state it is not compatible with non-relativistic particles (approximations were made assuming beta=1) but this is low priority.

What about Rad passmethods? I think there are constants hard coded there as well that assume beta=1, we should probably make sure these are correct in all cases...

There might be others too, so I agree we should be very careful with this kind of change

**Reply by lfarv** [reply]
> Introducing general `at` variables to define the particle type (starting with electrons only) that all function refer to would be a first step in this direction. This would also be coherent with the suggestion made in #285 to introduce a `Constants.py` module to centralized physics constants.

In fact, I already have this done in the `particle_data` branch. The `constants.py` module is there, there is a `Particle` object, defined by name, rest mass and charge, which is set as an attribute of the `Lattice` object. This is the easy part, but it's useless until the tracking works. So I'm now thinking of going the other way: set up the tracking first, still restricted to electrons, and then step by step, introduce the `Particle` object, move all hard-coded references to it, and finally allow other particles than electrons.

**Comment by swhite2401** [comment]
Sounds very good! Let me know in case help is needed for testing etc...
